README.MD
<p><img src="http://tailpress.io/images/tailpress100.svg" width="260" alt="SMC"></p>

[![GitHub release](https://img.shields.io/github/release/jeffreyvr/tailpress?include_prereleases=&sort=semver)](https://github.com/jeffreyvr/tailpress/releases/) [![License](https://img.shields.io/badge/License-MIT-blue)](#license) [![issues - tailpress](https://img.shields.io/github/issues/jeffreyvr/tailpress)](https://github.com/jeffreyvr/tailpress/issues)

# Introduction

SMC is a minimal boilerplate theme for WordPress using [Tailwind CSS](https://tailwindcss.com/).

## Getting started

### Using the installer

You can get started using the installer (using composer):

```bash
composer global require jeffreyvanrossum/tailpress-installer

tailpress new example-theme
```

_If you haven't already, make sure to place the `~/.composer/vendor/bin` directory in your `PATH` so the tailpress executable is found when you run the tailpress command in your terminal._

You can optionally set the theme name.

```bash
tailpress new example-theme --name="Example Theme"
```

By default, SMC uses Laravel Mix for compiling. Rather use Esbuild?

```bash
tailpress new example-theme --compiler="esbuild"
```

You can also initialize a new Git repository (branch defaults to `main`):

```bash
tailpress new example-theme --name="Example Theme" --git --branch="main"
```

Once your theme is ready, don't forget to cd into the directory.

You will be asked if you would like to have WordPress installed as well. Keep in mind that you still need a local development environment for PHP and MySQL.

### Regular method

- Clone repo `git clone https://github.com/jeffreyvr/tailpress.git && cd tailpress`
- Run `rm -rf .git` to remove git (or `rmdir .git` for Windows)
- Run `npm install`
- Run `npm run watch` to start developing

### General

You will find the editable CSS and Javascript files within the `/resources` folder.

Before you use your theme in production, make sure you run `npm run production`.

## NPM Scripts

There are several NPM scripts available. You'll find the full list in the `package.json` file under "scripts". A script is executed through the terminal by running `npm run script-name`.

| Script | Description |
| --- | --- |
| production | Creates a production (minified) build of app.js, app.css and editor-style.css. |
| dev | Creates a development build of app.js, app.css and editor-style.css. |
| watch | Runs several watch scripts concurrently. |

## Block editor support

SMC comes with support for the [block editor](https://wordpress.org/support/article/wordpress-editor/).

A basic setup for `theme.json` is included. This also means that you need to at least use WordPress 5.8. If you wan't to support earlier WordPress versions, you can use an [older version](https://github.com/jeffreyvr/tailpress/tree/0.1.1) of SMC instead.

CSS-classes for alignment (full, wide etc.) are generated automatically. You can opt-out on this by removing the plugin from the `tailwind.config.js` file.

To make the editing experience within the block editor more in line with the front end styling, a `editor-style.css` is generated.

### Define theme colors and font sizes

Several colors and font sizes are defined from the beginning. You can modify them in `theme.json`.

## Links

- [SMC website](https://tailpress.io)
- [Screencasts](https://www.youtube.com/playlist?list=PL6GBdOp044SHIOSCZejodwr1HcYsC43wG)
- [Tailwind CSS Documentation](https://tailwindcss.com/docs)
- [Laravel Mix Documentation](https://laravel-mix.com)
- [Esbuild Documentation](https://esbuild.github.io)

## Contributors

- [Jeffrey van Rossum](https://github.com/jeffreyvr)
- [All contributors](https://github.com/jeffreyvr/tailpress/graphs/contributors)

## License

MIT. Please see the [License File](/LICENSE) for more information.


CHANGELOG.md
# Changelog

All notable changes to SMC will be documented in this file.

## Unreleased

- Use `mix.options({ manifest: false })` instead of deprecated `Mix.manifest.refresh = _ => void 0`

## 3.3.0

- Update to Tailwind 3.3.0
- Laravel Mix is now the default compiler (with the SMC installer (^v2.0.0), use `compiler="esbuild"` if you want to keep using esbuild)

## 3.2.0

- Update to Tailwind 3.2.0

## 3.1.0

- Tailwind font sizes are now set as defined in `theme.json`.
- Breakpoints now based on WordPress defaults (https://developer.wordpress.org/block-editor/reference-guides/packages/packages-viewport/#usage).
- Providing `w-content`, `max-w-content`, `w-wide` and `max-w-wide` utility classes.
- Content width is now actually the width as defined in `theme.json`.
- Fixing align wide, width as defined in `theme.json`.
- Updating Tailwind CSS to version 3.1.0.
- Fix issues package.json scripts on Windows.

## 3.0.0 - 2021-12-14

- Updating Tailwind to 3.0.0.

### SMC installer

- The SMC installer (^0.2.0) now allows you to use Laravel Mix instead of esbuild by setting --compiler=mix.
- You may now also set dbname, dbuser, dbpass and dbhost.

## 2.0.0 - 2021-09-03

- Switching to Tailwind CLI and esbuild instead of LaraveL Mix.
- Removing `theme` subdirectory setup as it is no longer needed with the new build setup.
- Removing `SMC` class and it's functions (`tailpress()->get_header()` etc.) throughout the theme.
- New `tailpress_asset` function to get the URL of an asset (previously `tailpress_mix`).
- `tailpress_asset` function thaty appends a `time` parameter if [wp_get_environment_type()](https://developer.wordpress.org/reference/functions/wp_get_environment_type/) does not return `production` for cache busting (instead of the previously used versioned assets through `mix-manifest.json`).
- Update screenshot.png.
- Remove `block-editor.css`, only use `editor-style.css`.
- Moving `editor-style.css` from root to `css` directory.
- Update readme.

## 1.0.0 - 2021-08-25

- Replace `tailpress.json` with `theme.json` as used by WordPress core.
- Move template files into `theme` subdirectory.
- Move tailwind plugin to a [separate repository](https://github.com/jeffreyvr/tailwindcss-tailpress).
- Update readme and adding section on using installer.

## 0.1.0 - 2021-06-17

- No longer depending on jQuery.
- Fixes text color classes for the Block Editor.
- Use safelist.txt to prevent WP classes from being purged.
- Readme changes.
- MIT License.

## 0.0.9 - 2021-04-05

- Updating to Tailwind CSS v2.1 which includes the JIT engine in core among other things.

## 0.0.8 - 2021-03-23

- Using TailwindCSS JIT for way faster compiling.
- Updated readme.
- Fix loading styling in block editor.
- Check if mix-manifest.json file exists to prevent warning message.

## 0.0.7 - 2021-02-15

- Adding the option to apply submenu_class to the wp_nav_menu args.
- Adding the option to apply classes on li_class and submenu_class on specific depths, like: li_class_0.

### 0.0.6 - 2021-02-08

- Fixes issue on Windows.

## 0.0.5 - 2020-12-24

- Set selectors on single line since this seems to cause issues (nested CSS) with production build (#241a612).

## 0.0.4 - 2020-12-23

- Add nested CSS support for PostCSS.
- Minor readme changes.

## 0.0.3 - 2020-12-22

- Update Laravel Mix from version 5^ to 6^.
- Removing Laravel Mix Tailwind, defining plugins within webpack.mix.js instead.
- Switching from Sass to PostCSS for faster compiling.
- Moved SMC colors and font size settings to tailpress.json file.
- Use tailpress.json to populate editor-color-palette and editor-font-sizes theme support automatically.
- New screenshot.
- Update readme.
- Other minor fixes and improvements.

## 0.0.2 - 2020-11-24

- Adding basic support for the block editor Gutenberg by generating alignment, font size and color classes. Contains four theme colors out of the box, being primary, secondary, dark and light. This is adjustable of course.
- Loading a editor-style.css.
- Removing double slashes on resulting manifest asset URLs.
- Modified template files to have a better starting point (including horizontal main navigation, footer always at the bottom for short pages).
- Added a basic 404 page template.

## 0.0.1 - 2020-11-19

- Init release.


header.php
<!DOCTYPE html>
<html <?php language_attributes(); ?>>

<head>
  <meta charset="<?php bloginfo('charset'); ?>">
  <meta name="viewport" content="width=device-width">
  <link rel="profile" href="http://gmpg.org/xfn/11">
  <link rel="pingback" href="<?php bloginfo('pingback_url'); ?>">

  <?php wp_head(); ?>
</head>

<body <?php body_class('bg-white text-gray-900 antialiased'); ?>>

  <?php do_action('smc_site_before'); ?>

  <div id="page" class="min-h-screen flex flex-col">

    <?php do_action('smc_header'); ?>

    <header>

      <div class="mx-auto container">
        <div class="lg:flex lg:justify-between lg:items-center border-b py-6">
          <div class="flex justify-between items-center">
            <div>
              <?php if (has_custom_logo()) { ?>
                <?php the_custom_logo(); ?>
              <?php } else { ?>
                <a href="<?php echo get_bloginfo('url'); ?>" class="font-extrabold text-lg uppercase">
                  <?php echo get_bloginfo('name'); ?>
                </a>

                <p class="text-sm font-light text-gray-600">
                  <?php echo get_bloginfo('description'); ?>
                </p>

              <?php } ?>
            </div>

            <div class="lg:hidden">
              <a href="#" aria-label="Toggle navigation" id="primary-menu-toggle">
                <svg viewBox="0 0 20 20" class="inline-block w-6 h-6" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
                  <g stroke="none" stroke-width="1" fill="currentColor" fill-rule="evenodd">
                    <g id="icon-shape">
                      <path d="M0,3 L20,3 L20,5 L0,5 L0,3 Z M0,9 L20,9 L20,11 L0,11 L0,9 Z M0,15 L20,15 L20,17 L0,17 L0,15 Z" id="Combined-Shape"></path>
                    </g>
                  </g>
                </svg>
              </a>
            </div>
          </div>

          <?php
          wp_nav_menu(
            array(
              'container_id'    => 'primary-menu',
              'container_class' => 'hidden bg-gray-100 mt-4 p-4 lg:mt-0 lg:p-0 lg:bg-transparent lg:block',
              'menu_class'      => 'lg:flex lg:-mx-4',
              'theme_location'  => 'primary',
              'li_class'        => 'lg:mx-4',
              'fallback_cb'     => false,
            )
          );
          ?>
        </div>
      </div>
    </header>

    <div id="content" class="site-content flex-grow">

      <?php if (is_front_page()) { ?>
        <!-- Start introduction -->
        <div class="container mx-auto">
          <div class="px-12 py-16 my-12 rounded-xl bg-gradient-to-r from-blue-50 from-10% via-sky-100 via-30% to-blue-200 to-90%">
            <div class="mx-auto max-w-screen-md">
              <h1 class="text-3xl lg:text-6xl tracking-tight font-extrabold text-gray-800 mb-6">Start building your next <a href="https://tailwindcss.com" class="text-secondary">Tailwind CSS</a> flavoured WordPress theme
                with <a href="https://smc.io" class="text-primary">SMC</a>.</h1>
              <p class="text-gray-600 text-xl font-medium mb-10">SMC is your go-to starting
                point for developing WordPress themes with Tailwind CSS and comes with basic block-editor support out
                of the box.</p>
              <a href="https://github.com/jeffreyvr/smc" class="w-full sm:w-auto flex-none bg-gray-900 text-white text-lg leading-6 font-semibold py-3 px-6 border border-transparent rounded-xl focus:ring-2 focus:ring-offset-2 focus:ring-offset-white focus:ring-gray-900 focus:outline-none transition-colors duration-200">View
                on GitHub</a>
            </div>
          </div>
        </div>
        <!-- End introduction -->
      <?php } ?>

      <?php do_action('smc_content_start'); ?>

      <main>